test/goaway_test.go

```diff
@@ -0,0 +1,74 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package test
+
+import (
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+	testpb "google.golang.org/grpc/test/grpc_testing"
+)
+
+// TestGracefulClientOnGoAway attempts to ensure that when the server sends a
+// goaway, a client will never see an error.  This requires that the client is
+// appraised of the GOAWAY and updates its state accordingly before the
+// transport stops accepting new streams.  If a subconn is chosen by a picker
+// and receives the goaway before creating the stream, an error will occur, but
+// upon transparent retry, the clientconn will ensure a ready subconn is
+// chosen.
```

Comment:
The comment says what this test is for, but doesn't say how.
Mention that GOAWAY is triggered by max connection age? Either here or in the test body.